Create konnector_declaration_v3.md

doubleface opened the pull request:
Here is a proposition of declaration of konnectors to the new cozy stack (Gozy, v3)

nicofrand commented:
What are the verbs in the permissions ? Why should a konnector developer bother with those ?

ZeHiro commented:
I agree with @nicofrand : it's these verbs are not really meaningful. read/write is clearer.

nicofrand commented:
Do you have an example of a konnector that would only need a read access ? Even read/write does not make sense to me.

ZeHiro commented:
In fact it might be read data from provider or write data to the provider

doubleface commented:
@ZeHiro about the icons, I supposed it would be in the client application but it may be better to put it and declare it in the konnectors, now that it has its own repository. We "just" have to find a good protocol to serve it to the client. I will look at it.

ZeHiro commented:
@doubleface if each konnector has its own repo/package, everything should be bundled in the connector. 
@Phyks in its project to integrate weboob to konnectors used based64 images passed throw json API

doubleface commented:
@ZeHiro @nicofrand About the permission. These are permissions of access to given docTypes. I just used exactly the same permissions as the ones for the applications. You can have more information [in the official documentation](https://cozy.github.io/cozy-stack/permissions.html).

And no, I did not find any existing konnector which would only need read access by it still may be possible. I am working in my spare time on a RSS feeds konnector which would need read access to a list of feeds and read/write access to the list of items. The list of feeds would be updated by a specific application.

But if this type of permissions make contribution more complicated, it may be better to propose some shortcuts like : 

```
"permissions" : ["io.cozy.bills", "io.cozy.files"]
```
Which give all permissions on theses docTypes.
Does it make more sense?

nicofrand commented:
Yup, the shortcuts would make more sense to me.

ZeHiro commented:
I would say, as long as the permissions are really clear, I don't really care :)

nono commented on docs/konnector_declaration_v3.md:
```diff
@@ -0,0 +1,63 @@
+  and not the whole package (like the applications).
+- We want to propose a curated list of konnectors, but we also want to let the user install konnectors from other 
+  sources (github, npm, etc)
+- The konnectors will be executed in a container with the last LTS nodejs version (6 at the moment) with cozy-client-js giving access to the cozy stack and with access to internet (of course!)
```
Node 8 will be released in april. I think we will start with it

nono commented on docs/konnector_declaration_v3.md:
```diff
@@ -0,0 +1,63 @@
+
+## Proposition
+
+ - The konnectors will be installed using npm. Any source usable by npm is possible : github, gist, gitlab, bitbucket (git) 
```
npm or yarn? I have a preference for yarn, but it may be more difficult to use (no global install for example)

doubleface replied:
I am not a regular yarn user. yarn.lock is quite fine and we won't need global install from what I know, but ```yarn add``` does not allow install from gitlab or bitbucket (but url of tar.gz files allowed).
Maybe yarn will be better since it is also faster.

nono commented on docs/konnector_declaration_v3.md:
```diff
@@ -0,0 +1,63 @@
+ - The konnectors will be installed using npm. Any source usable by npm is possible : github, gist, gitlab, bitbucket (git) 
+   or a tar.gz somewhere on the internet (but not file addresse). More information here : https://docs.npmjs.com/cli/install
+ - A konnector package will need the following items :
+    - An index.js which will be run by the stack in a securised container
```
We can use the `main` field from `package.json` and fallback to `index.js` if missing

nono replied:
Cf https://docs.npmjs.com/files/package.json#main

doubleface replied:
OK, will update it

nono commented on docs/konnector_declaration_v3.md:
```diff
@@ -0,0 +1,63 @@
+        "description": "Required to save the bills pdf",
+        "type": "io.cozy.files",
+        "verbs": ["GET", "POST"]
+      },
```
Can you fix indentation here?

doubleface replied:
done

nono commented on docs/konnector_declaration_v3.md:
```diff
@@ -0,0 +1,63 @@
+    "source": "https://github.com/cozy/trainline@build",
+    "locales": {
+        "fr": {
+            description: "Récupère toutes vos factures trainline"
```
Can you add the translation of the description of permissions?

doubleface replied:
done

nono commented:
`verbs` is optional. But we want to have in the future some konnectors that can synchronize things from the cozy to external services and a read-only permission (`GET`) makes sense there.

We also think it's nice to have a description for each permission (that can be localized) to explain to the user why the permission is needed.

nicofrand commented:
If a konnector is now able to write things toward the provider, I'd prefer that the user allows it instead of declaring it in the konnector.

nono commented:
> If a konnector is now able to write things toward the provider, I'd prefer that the user allows it instead of declaring it in the konnector.

It's the goal. The user must accept the permissions to use the konnectors.

ZeHiro commented:
@nono how this will be implemented konnector-side ? If the konnector "wants" to push information to a provider, it will just do it. Will the data be pushed by the plateform and not the konnector ?

doubleface commented:
@nicofrand @ZeHiro after @nono remark about permissions, I don't think the shortcut I proposed is OK. Since the verbs are optional, I think this is easy enough and we still need to define the text which will be displayed to the user to accept the permissions. So I will remove the "verbs" attribute in my proposition.

doubleface commented:
@nicofrand @ZeHiro I updated the document with your remarks and also removed npm and use git only instead. Npm could not do version management with git repositories and it is easier to publish on github than on npm (for the majority of users I think).

If you are OK, I will merge this document and publish a post on the forum to introduce it and have a wider range of feedback. Anyway, this document is not definitive at all at the moment.

nono commented on docs/konnector_declaration_v3.md:
```diff
@@ -28,6 +28,7 @@ remarks are welcome.
 {
   "version": "1.0.0",
     "name": "Trainline",
+    "type": "node",
```
Can you also add the `main` in this example?

doubleface replied:
Actually, I supposed the ```main``` section would be in the package.json but that way, we would not have to read the package.json at all and we would run the konnector with a simple node command not using npm. Ok, I put the main section in the manifest.

nono replied:
Sorry, my bad, you are right, it should be in package.json

doubleface replied:
It is not so bad if we don't want to have to read in the package.json to run the konnector. Or we can run the konnector with npm start and let the developper of the konnector handle that part.
